### ISSUE TITLE
CMake can't find required information from nvcc

### ISSUE THREAD

**Issue by dbeurle** [negativo17/cuda#9]
CMake looks for some particular flags from `nvcc -v *.cu` to setup the required variables.  However they cannot be found and the configuration stage fails.  Is this something on this package side or is this set by NVIDIA?  I'm running the latest package on Fedora 27.

More information in the issue on CMake GitLab:
https://gitlab.kitware.com/cmake/cmake/issues/17776#note_381515

**Comment by scaronni** [comment]
Mmm. Where should they come from?

```
$ sudo HOST_COMPILER=cuda-g++ make 
/usr/bin/nvcc -v --include-path /usr/include/cuda -ccbin cuda-g++ -I../../common/inc  -m64    -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_52,code=sm_52 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_70,code=sm_70 -gencode arch=compute_70,code=compute_70 -o deviceQuery.o -c deviceQuery.cpp
#$ _SPACE_= 
#$ _CUDART_=cudart
#$ _HERE_=/usr/bin
#$ _THERE_=/usr/bin
#$ _TARGET_SIZE_=
#$ _TARGET_DIR_=
#$ _TARGET_SIZE_=64
#$ cuda-g++ -std=c++14 -c -x c++ -D__NVCC__  -I"/usr/include/cuda" -I"../../common/inc" -D"__CUDACC_VER_BUILD__=85" -D"__CUDACC_VER_MINOR__=1" -D"__CUDACC_VER_MAJOR__=9" -m64 -o "deviceQuery.o" "deviceQuery.cpp" 
/usr/bin/nvcc -v --include-path /usr/include/cuda -ccbin cuda-g++   -m64      -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_52,code=sm_52 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_70,code=sm_70 -gencode arch=compute_70,code=compute_70 -o deviceQuery deviceQuery.o 
#$ _SPACE_= 
#$ _CUDART_=cudart
#$ _HERE_=/usr/bin
#$ _THERE_=/usr/bin
#$ _TARGET_SIZE_=
#$ _TARGET_DIR_=
#$ _TARGET_SIZE_=64
#$ nvlink --arch=sm_30 --register-link-binaries="/tmp/tmpxft_00004602_00000000-2_deviceQuery_dlink.reg.c"  -m64 -cpu-arch=X86_64 -report-arch "deviceQuery.o"  -o "/tmp/tmpxft_00004602_00000000-4_deviceQuery_dlink.compute_30.cubin"
#$ nvlink --arch=sm_35 --register-link-binaries="/tmp/tmpxft_00004602_00000000-2_deviceQuery_dlink.reg.c"  -m64 -cpu-arch=X86_64 -report-arch "deviceQuery.o"  -lcudadevrt  -o "/tmp/tmpxft_00004602_00000000-5_deviceQuery_dlink.compute_35.cubin"
#$ nvlink --arch=sm_37 --register-link-binaries="/tmp/tmpxft_00004602_00000000-2_deviceQuery_dlink.reg.c"  -m64 -cpu-arch=X86_64 -report-arch "deviceQuery.o"  -lcudadevrt  -o "/tmp/tmpxft_00004602_00000000-6_deviceQuery_dlink.compute_37.cubin"
#$ nvlink --arch=sm_50 --register-link-binaries="/tmp/tmpxft_00004602_00000000-2_deviceQuery_dlink.reg.c"  -m64 -cpu-arch=X86_64 -report-arch "deviceQuery.o"  -lcudadevrt  -o "/tmp/tmpxft_00004602_00000000-7_deviceQuery_dlink.compute_50.cubin"
#$ nvlink --arch=sm_52 --register-link-binaries="/tmp/tmpxft_00004602_00000000-2_deviceQuery_dlink.reg.c"  -m64 -cpu-arch=X86_64 -report-arch "deviceQuery.o"  -lcudadevrt  -o "/tmp/tmpxft_00004602_00000000-8_deviceQuery_dlink.compute_52.cubin"
#$ nvlink --arch=sm_60 --register-link-binaries="/tmp/tmpxft_00004602_00000000-2_deviceQuery_dlink.reg.c"  -m64 -cpu-arch=X86_64 -report-arch "deviceQuery.o"  -lcudadevrt  -o "/tmp/tmpxft_00004602_00000000-9_deviceQuery_dlink.compute_60.cubin"
#$ nvlink --arch=sm_61 --register-link-binaries="/tmp/tmpxft_00004602_00000000-2_deviceQuery_dlink.reg.c"  -m64 -cpu-arch=X86_64 -report-arch "deviceQuery.o"  -lcudadevrt  -o "/tmp/tmpxft_00004602_00000000-10_deviceQuery_dlink.compute_61.cubin"
#$ nvlink --arch=sm_70 --register-link-binaries="/tmp/tmpxft_00004602_00000000-2_deviceQuery_dlink.reg.c"  -m64 -cpu-arch=X86_64 -report-arch "deviceQuery.o"  -lcudadevrt  -o "/tmp/tmpxft_00004602_00000000-11_deviceQuery_dlink.compute_70.sm_70.cubin"
#$ fatbinary --create="/tmp/tmpxft_00004602_00000000-3_deviceQuery_dlink.fatbin" -64 --cmdline="" -link "--image=profile=sm_30@compute_30,file=/tmp/tmpxft_00004602_00000000-4_deviceQuery_dlink.compute_30.cubin" "--image=profile=sm_35@compute_35,file=/tmp/tmpxft_00004602_00000000-5_deviceQuery_dlink.compute_35.cubin" "--image=profile=sm_37@compute_37,file=/tmp/tmpxft_00004602_00000000-6_deviceQuery_dlink.compute_37.cubin" "--image=profile=sm_50@compute_50,file=/tmp/tmpxft_00004602_00000000-7_deviceQuery_dlink.compute_50.cubin" "--image=profile=sm_52@compute_52,file=/tmp/tmpxft_00004602_00000000-8_deviceQuery_dlink.compute_52.cubin" "--image=profile=sm_60@compute_60,file=/tmp/tmpxft_00004602_00000000-9_deviceQuery_dlink.compute_60.cubin" "--image=profile=sm_61@compute_61,file=/tmp/tmpxft_00004602_00000000-10_deviceQuery_dlink.compute_61.cubin" "--image=profile=sm_70@compute_70,file=/tmp/tmpxft_00004602_00000000-11_deviceQuery_dlink.compute_70.sm_70.cubin" --embedded-fatbin="/tmp/tmpxft_00004602_00000000-3_deviceQuery_dlink.fatbin.c" 
#$ rm /tmp/tmpxft_00004602_00000000-3_deviceQuery_dlink.fatbin
#$ cuda-g++ -std=c++14 -c -x c++ -DFATBINFILE="\"/tmp/tmpxft_00004602_00000000-3_deviceQuery_dlink.fatbin.c\"" -DREGISTERLINKBINARYFILE="\"/tmp/tmpxft_00004602_00000000-2_deviceQuery_dlink.reg.c\"" -I. -D__NV_EXTRA_INITIALIZATION= -D__NV_EXTRA_FINALIZATION=  -I"/usr/include/cuda" -D"__CUDACC_VER_BUILD__=85" -D"__CUDACC_VER_MINOR__=1" -D"__CUDACC_VER_MAJOR__=9" -m64 -o "/tmp/tmpxft_00004602_00000000-12_deviceQuery_dlink.o" "/usr/bin/crt/link.stub" 
#$ cuda-g++ -m64 -o "deviceQuery" -std=c++14 -Wl,--start-group "/tmp/tmpxft_00004602_00000000-12_deviceQuery_dlink.o" "deviceQuery.o"  -lcudadevrt  -lcudart_static  -lrt -lpthread  -ldl  -Wl,--end-group 
mkdir -p ../../bin/x86_64/linux/release
cp deviceQuery ../../bin/x86_64/linux/release
```

**Comment by scaronni** [comment]
Maybe it prints them if they are set in the first place?

But `NVVMIR_LIBRARY_DIR` is set, and is not printed anyway:

```
$ cat /etc/profile.d/cuda.sh 
[ -x /usr/bin/nvcc ] && export NVVMIR_LIBRARY_DIR=/usr/share/cuda
[ -x /usr/libexec/cuda/open64/bin/nvopencc ] && export PATH=$PATH:/usr/libexec/cuda/open64/bin
[ -d /usr/include/cuda ] && export CUDA_INC_PATH=/usr/include/cuda
```

**Comment by invexed** [comment]
I too am having this problem on CentOS 7.

**Comment by scaronni** [comment]
I guess there is some variable that needs to be set for CMake, don't know which.

**Comment by scaronni** [comment]
https://cmake.org/cmake/help/latest/module/FindCUDA.html

**Comment by dbeurle** [comment]
The RPMs provided by NVIDIA (https://rpmfusion.org/Howto/CUDA) have the required PATH variables set when compiling with `nvcc -v *.cu`.  If I use find_package(CUDA) it finds the runtime fine.  The problem occurs when CMake tests `nvcc` and it doesn't have the required environment variables.

**Comment by dbeurle** [comment]
The problem is with a CMakeLists.txt when specifying the languages option
```cmake
project(my_project VERSION 1.0 LANGUAGES CXX CUDA)
```

Those variables are printing when using the RPMFusion `nvcc`:

```
#$ _SPACE_= 
#$ _CUDART_=cudart
#$ _HERE_=/usr/local/cuda-9.1/bin
#$ _THERE_=/usr/local/cuda-9.1/bin
#$ _TARGET_SIZE_=
#$ _TARGET_DIR_=
#$ _TARGET_DIR_=targets/x86_64-linux
#$ TOP=/usr/local/cuda-9.1/bin/..
#$ NVVMIR_LIBRARY_DIR=/usr/local/cuda-9.1/bin/../nvvm/libdevice
#$ LD_LIBRARY_PATH=/usr/local/cuda-9.1/bin/../lib:
#$ PATH=/usr/local/cuda-9.1/bin/../nvvm/bin:/usr/local/cuda-9.1/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin
#$ INCLUDES="-I/usr/local/cuda-9.1/bin/../targets/x86_64-linux/include"  
#$ LIBRARIES=  "-L/usr/local/cuda-9.1/bin/../targets/x86_64-linux/lib/stubs" "-L/usr/local/cuda-9.1/bin/../targets/x86_64-linux/lib"
#$ CUDAFE_FLAGS=
#$ PTXAS_FLAGS=
#$ gcc -std=c++14 -D__CUDA_ARCH__=300 -E -x c++  -DCUDA_DOUBLE_MATH_FUNCTIONS -D__CUDACC__ -D__NVCC__  "-I/usr/local/cuda-9.1/bin/../targets/x86_64-linux/include"   -D"__CUDACC_VER_BUILD__=85" -D"__CUDACC_VER_MINOR__=1" -D"__CUDACC_VER_MAJOR__=9" -include "cuda_runtime.h" -m64 "cuda_ptx_test.cu" > "/tmp/tmpxft_00000fc8_00000000-8_cuda_ptx_test.cpp1.ii" 
#$ cicc --c++14 --gnu_version=70201 --allow_managed   -arch compute_30 -m64 -ftz=0 -prec_div=1 -prec_sqrt=1 -fmad=1 --include_file_name "tmpxft_00000fc8_00000000-2_cuda_ptx_test.fatbin.c" -tused -nvvmir-library "/usr/local/cuda-9.1/bin/../nvvm/libdevice/libdevice.10.bc" --gen_module_id_file --module_id_file_name "/tmp/tmpxft_00000fc8_00000000-3_cuda_ptx_test.module_id" --orig_src_file_name "cuda_ptx_test.cu" --gen_c_file_name "/tmp/tmpxft_00000fc8_00000000-5_cuda_ptx_test.cudafe1.c" --stub_file_name "/tmp/tmpxft_00000fc8_00000000-5_cuda_ptx_test.cudafe1.stub.c" --gen_device_file_name "/tmp/tmpxft_00000fc8_00000000-5_cuda_ptx_test.cudafe1.gpu"  "/tmp/tmpxft_00000fc8_00000000-8_cuda_ptx_test.cpp1.ii" -o "/tmp/tmpxft_00000fc8_00000000-5_cuda_ptx_test.ptx"

```
which works as expected with CMake.

**Comment by scaronni** [comment]
Thanks, I will check what is missing in the packages compared to the Nvidia provided ones.

**Comment by scaronni** [comment]
Ok, figured out, those are coming from the `nvcc.profile` that is sitting in the same directory as the `nvcc` binary. I thought it was not needed (didn't hit this so far, as what's the point of declaring empty variables?) as it uses the same system path as the rest.

**Comment by dbeurle** [comment]
Thanks for that - much appreciated.  Will this hit the repo shortly?

**Comment by scaronni** [comment]
Yes, in a few hours. But if you prefer to use the official CUDA packages + RPMFusion and just need a compiler yoI'm building now just `cuda-gcc` here: https://copr.fedorainfracloud.org/coprs/slaanesh/cuda-gcc/

**Comment by dbeurle** [comment]
Perfect!  Personally, I prefer these packages to the official ones, they're just cleaner.  Thanks again for the work.

**Comment by scaronni** [comment]
I still think the correct way would be for CMake to read environment variables and not the debugging output of `nvcc`. Bah.

Anyway, can you please make a test by populating `/usr/bin/nvcc.profile` with this content:
```
NVVMIR_LIBRARY_DIR = /usr/share/cuda
PATH += /usr/libexec/cuda/open64/bin:
INCLUDES += "-I/usr/include/cuda "
LIBRARIES =+ " -L/usr/lib64"
CUDAFE_FLAGS +=
PTXAS_FLAGS +=
```

If it works, then please try to remove the `LIBRARIES` line. If it works I will put the one without the `LIBRARIES` line and remove the environment variables in `/etc/profile.d/cuda.*`.

Thanks.

**Comment by scaronni** [comment]
I still think the correct way would be for CMake to read environment variables and not the debugging output of `nvcc`. Bah.

Anyway, can you please make a test by populating `/usr/bin/nvcc.profile` with this content:
```
NVVMIR_LIBRARY_DIR = /usr/share/cuda
PATH += /usr/libexec/cuda/open64/bin:
INCLUDES += "-I/usr/include/cuda "
LIBRARIES =+ " -L/usr/lib64"
CUDAFE_FLAGS +=
PTXAS_FLAGS +=
```

If it works, then please try to remove the `LIBRARIES` line. If it works I will put the one without the `LIBRARIES` line and remove the environment variables in `/etc/profile.d/cuda.*`.

Thanks.

**Comment by dbeurle** [comment]
Sure I'll give that a shot.  Might take a while as I'll need to purge the RPMFusion packages and reinstall your's.

You can suggest that on the issue I linked to.  I find doing anything with CUDA far too fragile.

**Comment by scaronni** [comment]
It's fragile because it's not open source and tries to accomodate all distributions unfortunately.

**Comment by dbeurle** [comment]
```
$ cmake ..
-- The CXX compiler identification is GNU 7.2.1
-- The CUDA compiler identification is unknown
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working CUDA compiler: /usr/bin/nvcc
-- Check for working CUDA compiler: /usr/bin/nvcc -- broken
CMake Error at /usr/share/cmake/Modules/CMakeTestCUDACompiler.cmake:46 (message):
  The CUDA compiler

    "/usr/bin/nvcc"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/dbeurle/apps/cuda_test_project/build/CMakeFiles/CMakeTmp
    
    Run Build Command:"/usr/bin/gmake" "cmTC_1f5bf/fast"
    /usr/bin/gmake -f CMakeFiles/cmTC_1f5bf.dir/build.make CMakeFiles/cmTC_1f5bf.dir/build
    gmake[1]: Entering directory '/home/dbeurle/apps/cuda_test_project/build/CMakeFiles/CMakeTmp'
    Building CUDA object CMakeFiles/cmTC_1f5bf.dir/main.cu.o
    /usr/bin/nvcc     -x cu -c /home/dbeurle/apps/cuda_test_project/build/CMakeFiles/CMakeTmp/main.cu -o CMakeFiles/cmTC_1f5bf.dir/main.cu.o
    Linking CUDA device code CMakeFiles/cmTC_1f5bf.dir/cmake_device_link.o
    /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_1f5bf.dir/dlink.txt --verbose=1
    /usr/bin/nvcc       -shared -dlink CMakeFiles/cmTC_1f5bf.dir/main.cu.o -o CMakeFiles/cmTC_1f5bf.dir/cmake_device_link.o 
    nvlink fatal   : Could not open input file ' -L/usr/lib64'
    gmake[1]: *** [CMakeFiles/cmTC_1f5bf.dir/build.make:98: CMakeFiles/cmTC_1f5bf.dir/cmake_device_link.o] Error 1
    gmake[1]: Leaving directory '/home/dbeurle/apps/cuda_test_project/build/CMakeFiles/CMakeTmp'
    gmake: *** [Makefile:126: cmTC_1f5bf/fast] Error 2
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:3 (project)


-- Configuring incomplete, errors occurred!
See also "/home/dbeurle/apps/cuda_test_project/build/CMakeFiles/CMakeOutput.log".
See also "/home/dbeurle/apps/cuda_test_project/build/CMakeFiles/CMakeError.log".
```

This error also occurs if I set
```
-DCMAKE_CXX_COMPILER=/usr/bin/cuda-g++ -DCUDA_HOST_COMPILER=/usr/bin/cuda-g++
```

**Comment by dbeurle** [comment]
Also - was this a typo?
```
LIBRARIES =+ " -L/usr/lib64"
```
to
```
LIBRARIES += " -L/usr/lib64"
```
?

It didn't change anything however.  If I remove the libraries line the path and the include seem to be set correctly in CMake.

**Comment by dbeurle** [comment]
Sorry for the spam.  CMake doesn't appreciate that space between the `"` and the `-L`.  Removing that fixes the problem and CMake configures correctly.

**Comment by scaronni** [comment]
Ah, good. Can you also remove the LIBRARIES line entirely? Does it work the same?

**Comment by dbeurle** [comment]
If I remove that LIBRARIES line we're back at square one.

**Comment by scaronni** [comment]
Ok, then libraries is.

Shall we also add the override for GCC to `cuda-gcc` in the file so on Fedora we can skip setting it on variables? In some places, like the examples, there's `-ccbin` explicitly set though.

**Comment by dbeurle** [comment]
It's a bit of a pain to set the CUDA host compiler every time a new project is configured but it's usually done once so it's not the end of the world.  Would it break the examples?  What does `-ccbin` currently point to?

**Comment by scaronni** [comment]
`-ccbin` points to `g++` in the various Makefiles for the samples, by default uses `g++`. So it might make more sense to be set. I'll come up with something tomorrow and ask you to test, if possible. Probably when installing on Fedora I'll set the dependency on `cuda-gcc` and an extra line `nvcc.profile`.

Many thanks.

**Comment by dbeurle** [comment]
Makes sense.  Happy to test.

**Comment by scaronni** [comment]
Makes sense?

https://github.com/negativo17/cuda/commit/cd2bda142241dfeaf0619681cfeb701d4f305d49

To test:

`/usr/bin/nvcc.profile` (same as above):
```
NVVMIR_LIBRARY_DIR = /usr/share/cuda
PATH += /usr/libexec/cuda/open64/bin:
INCLUDES += "-I/usr/include/cuda "
LIBRARIES =+ " -L/usr/lib64"
CUDAFE_FLAGS +=
PTXAS_FLAGS +=
```

`/etc/profile.d/cuda.sh`:
```
if [ -x /usr/bin/cuda-g++ ]; then
  export HOST_COMPILER=/usr/bin/cuda-g++
fi

if [ -x /usr/libexec/cuda/open64/bin/nvopencc ]
  export PATH=$PATH:/usr/libexec/cuda/open64/bin
fi
```
And `/etc/profile.d/cuda.csh`:
```
if ( -x /usr/bin/cuda-g++ ) then
  setenv HOST_COMPILER /usr/bin/cuda-g++
endif

if ( -x /usr/libexec/cuda/open64/bin/nvopencc ) then
  setenv PATH ${PATH}:/usr/libexec/cuda/open64/bin
endif
```

Logoff, login and you should be good to go. Then you can just do everywhere like in the samples, for example:
```
CUDA_PATH     ?= "/usr"
HOST_COMPILER ?= g++
NVCC          := $(CUDA_PATH)/bin/nvcc -ccbin $(HOST_COMPILER)
```
This way, you can just set the above in your project and it should just work if running it on a system with or without the `cuda-gcc-g++` package, with a cross compiler, etc. Also the includes are now found automatically and you don't need to specify them anymore as `'-I/usr/include/cuda'`. It should just work.

Please let me know if it works for you and I will push an update immediately.

Thanks!

**Comment by dbeurle** [comment]
Here's the kicker.  It seems you still need 
```bash
[ -d /usr/include/cuda ] && export CUDA_INC_PATH=/usr/include/cuda
```
in the environment script otherwise
```
CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find CUDA (missing: CUDA_INCLUDE_DIRS) (found suitable version
  "9.1", minimum required is "9.0")
```
:(

Also - I commented on a typo in the bash script, otherwise it fails on `source`.

**Comment by scaronni** [comment]
Mmm, from the documentation it seems to be `CUDA_INCLUDE_DIRS` and not `CUDA_INC_PATH`:

https://cmake.org/cmake/help/latest/module/FindCUDA.html

I will add `CUDA_INCLUDE_DIRS`.

> Also - I commented on a typo in the bash script, otherwise it fails on source.

Thanks, I will fix it.

I'm away from keyboard for a few hours. I'll trigger the build with the changes just a after.

**Comment by scaronni** [comment]
Packages are building with these commits:
https://github.com/negativo17/cuda/commits/master

**Comment by dbeurle** [comment]
I got those new packages.  There seems to be a regression.  Configuring with CMake gives
```
    Building CUDA object CMakeFiles/cmTC_94fb7.dir/main.cu.o
    /usr/bin/nvcc     -x cu -c /home/dbeurle/apps/neon/build/CMakeFiles/CMakeTmp/main.cu -o CMakeFiles/cmTC_94fb7.dir/main.cu.o
    error in open: /usr/bin/../nvvm/libdevice/libdevice.10.bc
    No such file or directory

```
It seems to find libdevice.10.bc which doesn't live here.  The nvcc.profile is incorrect here:
```
TOP              = $(_HERE_)/..

NVVMIR_LIBRARY_DIR = $(TOP)/nvvm/libdevice

LD_LIBRARY_PATH += $(TOP)/lib:
PATH            += $(TOP)/nvvm/bin:$(_HERE_):

INCLUDES        +=  "-I$(TOP)/$(_TARGET_DIR_)/include" $(_SPACE_)

LIBRARIES        =+ $(_SPACE_) "-L$(TOP)/$(_TARGET_DIR_)/lib$(_TARGET_SIZE_)/stubs" "-L$(TOP)/$(_TARGET_DIR_)/lib$(_TARGET_SIZE_)"

CUDAFE_FLAGS    +=
PTXAS_FLAGS     +=
```
where `NVVMIR_LIBRARY_DIR = $(TOP)/nvvm/libdevice` is the wrong directory - it should be `NVVMIR_LIBRARY_DIR = $(TOP)/usr/share/cuda`.  The nvcc.profile in the repo seems correct though.  Ideas?

**Comment by scaronni** [comment]
Hello, Saturday I pushed 1:9.1.85-3 with the above problems you mentioned because I missed that the `nvcc.profile` was being overwritten. It's fixed in 1:9.1.85-4, pushed saturday as well after a couple of hours.

**Comment by dbeurle** [comment]
Got it - works for me.  I'm happy to close this off if you are.

Thanks for the great help!

**Comment by robertmaynard** [comment]
From the CMake team, thanks for these fixes.

> I still think the correct way would be for CMake to read environment variables and not the debugging output of nvcc. Bah.

Background CMake wants the `LIBRARIES` and `INCLUDES` information so that when building mixed language applications we can use the C or C++ linker and propagate the correct implicit CUDA compiler information so that linking to CUDA SDK libraries works nicely.

I doubt we could have gotten this information from just environment variables without running nvcc, as we must support installations of nvcc in non standard locations, but we most likely could have reconstructed the paths for LIBRARIES and INCLUDES from TOP,TARGET_DIR, and TARGET_SIZE but that would CMake to redo work that  nvcc is currently doing. 

So again, thanks for restoring this information 👍

**Comment by scaronni** [comment]
Ok, thanks for the explanation and feedback!

Adding a couple of variables here and there is not a big deal as well.

**Comment by dbeurle** [comment]
Urgh, sorry to open this up again.  CMake doesn't like the `CUDA_INCLUDE_DIRS` and fails to find it.  Using the old `CUDA_INC_PATH` works.  Can you replicate this?

**Comment by scaronni** [comment]
Where do you see the need for `CUDA_INC_PATH`? the CMake module looks for `CUDA_INCLUDE_DIRS`:

https://cmake.org/cmake/help/latest/module/FindCUDA.html

**Comment by dbeurle** [comment]
Sorry, it was a red herring.  I just reset everything back to normal and logged out.  Seems there was something left over - all good now!

**Comment by scaronni** [comment]
Ok, thanks for feedback. Building now a new version with the 3 patches released so far for CUDA 9.1.
